File: examplesForThesis-R/ch4-example-gpo-garchinference.R
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GPO for ML inference in the GARCH(1,1) model
# Example 4.8 in Section 4.4.3
#
########################################################################

grid=seq(0.01,0.25,0.01)
toplott = c(3:7)

layout(matrix(1:15, 5, 3, byrow = FALSE))  
par(mar=c(4,5,1,3.5)) 

########################################################################
# PI rule
########################################################################

d <- read.table("ch4-example-gpo-garchinference-pi.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
d2 <- read.table("ch4-example-gpo-garchinference-pi2.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
d2 = as.numeric(d2)
d2th = d2[1:11];
d2ll = d2[12:21];

M = d[,1:10];
L = d[,11:20];
U = d[,21:30];
A = d[,30:40];

for ( ii in toplott ) {
  plot(grid,M[,ii],type="l",lwd=3,col="darkgreen",ylim=c(-300,-210),xlab=expression(alpha),ylab="Log-likelihood",xlim=c(0,0.25))
  polygon( c(grid,rev(grid)), c(U[,ii], rev(L[,ii])), col="lightgrey", border=NA)
  lines(grid,M[,ii],lwd=3,col="darkgreen")
  points(d2th[1:ii-1],d2ll[1:ii-1],pch=19,cex=1.5)
  
  par(new=T)
  AA = -as.numeric( A[,ii] )
  plot(grid,AA,lwd=2,col="darkgreen",lty="dotted",axes=F,xlab="",ylab="", ylim=c(0,2*max(AA)),type="l")
  axis(4, ylim=c(0,2*max(AA)),lwd=1)
  mtext(4,text="PI",line=2,cex=0.65)
}

########################################################################
# EI rule
########################################################################

d <- read.table("ch4-example-gpo-garchinference-ei.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
d2 <- read.table("ch4-example-gpo-garchinference-ei2.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
d2 = as.numeric(d2)
d2th = d2[1:11];
d2ll = d2[12:21];

M = d[,1:10];
L = d[,11:20];
U = d[,21:30];
A = d[,30:40];

for ( ii in toplott ) {
  plot(grid,M[,ii],type="l",lwd=3,col="darkred",ylim=c(-240,-210),xlab=expression(alpha),ylab="Log-likelihood",xlim=c(0,0.25))
  polygon( c(grid,rev(grid)), c(U[,ii], rev(L[,ii])), col="lightgrey", border=NA)
  lines(grid,M[,ii],lwd=3,col="darkred")
  points(d2th[1:ii-1],d2ll[1:ii-1],pch=19,cex=1.5)
  
  par(new=T)
  AA = -as.numeric( A[,ii] )
  plot(grid,AA,lwd=2,col="darkred",lty="dotted",axes=F,xlab="",ylab="", ylim=c(0,2*max(AA)),type="l")
  axis(4, ylim=c(0,2*max(AA)),lwd=1)
  mtext(4,text="EI",line=2,cex=0.65)
}

########################################################################
# UCB rule
########################################################################

d <- read.table("ch4-example-gpo-garchinference-ucb.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
d2 <- read.table("ch4-example-gpo-garchinference-ucb2.csv",header=T,sep=",",stringsAsFactors=F)[,-1]
d2 = as.numeric(d2)
d2th = d2[1:11];
d2ll = d2[12:21];

M = d[,1:10];
L = d[,11:20];
U = d[,21:30];
A = d[,30:40];

for ( ii in toplott ) {
  plot(grid,M[,ii],type="l",lwd=3,col="steelblue",ylim=c(-300,-210),xlab=expression(alpha),ylab="Log-likelihood",xlim=c(0,0.25))
  polygon( c(grid,rev(grid)), c(U[,ii], rev(L[,ii])), col="lightgrey", border=NA)
  lines(grid,M[,ii],lwd=3,col="steelblue")
  points(d2th[1:ii-1],d2ll[1:ii-1],pch=19,cex=1.5)
  
  par(new=T)
  AA = -as.numeric( A[,ii] )
  plot(grid,AA,lwd=2,col="steelblue",lty="dotted",axes=F,xlab="",ylab="", ylim=c(1.3*max(AA),-150),type="l")
  axis(4, ylim=c(2*max(AA),0),lwd=1)
  mtext(4,text="UCB",line=2,cex=0.65)
}

########################################################################
# End of file
########################################################################

File: examplesForThesis-python/ch4-example-gpo-garchinference.py
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GPO for ML inference in the GARCH(1,1) model
# Example 4.8 in Section 4.4.3 
#
########################################################################

from smc import *
from classes import *
from helpers import *
import numpy as np
from DIRECT import solve
import GPy
from gpohelpers import *
import pandas

########################################################################
# Arrange the data structures
########################################################################
data             = stData();
smc              = smcSampler();
par              = stParameters();

########################################################################
# Setup the system
########################################################################
sys              = stSystemGARCH()
sys.version      = "standard"
sys.par          = np.zeros((4,1))
sys.par[0]       = 0.10;
sys.par[1]       = 0.80;
sys.par[2]       = 0.05;
sys.par[3]       = 0.30;
sys.T            = 250;

thSys            = stSystemGARCH()
thSys.T          = sys.T;
thSys.version    = sys.version;

# Generate some data
data.sample(sys, np.zeros(sys.T));

########################################################################
# Setup the parameters for the algorithm
########################################################################

par.fileprefix     = "garch"
par.nPars          = 1;
par.epsilonEI      = 0.01;
par.epsilonUCB     = 0.01;
par.preIter        = 2;
par.maxIter        = 10;

smc.nPart          = 100;
smc.resamplingType = "systematic";     # multinomial or systematic
smc.filterType     = "fullyadapted";   # kalman or bootstrap or fullyadapted

########################################################################
# Preliminaries
########################################################################

# Make a grid to evaluate the EI on
l      = np.array([0.01], dtype=np.float64)
u      = np.array([0.25], dtype=np.float64)

# Preallocate the information needed
th     = np.zeros((par.maxIter+1,par.nPars))
EI     = np.zeros((par.maxIter+1,1))
mumax  = np.zeros((par.maxIter,1))
pSamp  = np.zeros((par.maxIter,1))
thhat  = np.zeros((par.maxIter,par.nPars))
llmax  = np.zeros((par.maxIter,1))
grid  = np.arange(0.01,0.26,0.01);
gridM = np.zeros((len(grid),par.maxIter));
gridL = np.zeros((len(grid),par.maxIter));
gridU = np.zeros((len(grid),par.maxIter));
gridA = np.zeros((len(grid),par.maxIter));

#-----------------------------------------------------------------------
# Evaluate the log-likelihood in some random points (many) to get the hyperparameters
#-----------------------------------------------------------------------
tmpth = np.zeros((30,1));
tmpll = np.zeros((30,1));
for kk in range(0,30):
    tmpth[kk]     =  l + (u-l) * np.random.random(par.nPars)
    
    thSys.storeParameters((tmpth[kk], sys.par[1], sys.par[2], sys.par[3]),sys,par);
    smc.faPF(data, thSys, par)
    tmpll[kk] = smc.ll;
    print kk;

# Fit the hyperparameters of the kernel
kernel = GPy.kern.Matern52(input_dim=par.nPars) + GPy.kern.bias(input_dim=par.nPars)
m = GPy.models.GPRegression(tmpth,tmpll,kernel)
m.optimize()
m.optimize_restarts(num_restarts = 10)  

########################################################################
# GPO using the EI rule
########################################################################

# Make a preliminary run to estimate the log-likelihood
for kk in range(0,par.preIter):
    th[kk,:]     =  l + (u-l) * np.random.random(par.nPars)
    
    thSys.storeParameters((th[kk,:], sys.par[1], sys.par[2], sys.par[3]),sys,par);
    smc.faPF(data, thSys, par)
    pSamp[kk] = smc.ll;
    print kk;

#-----------------------------------------------------------------------
# Main GPO-loop
#-----------------------------------------------------------------------
for kk in range(par.preIter-1,par.maxIter):

    # Sample the log-likelihood
    thP = (th[kk,:], sys.par[1], sys.par[2], sys.par[3]);
    thSys.storeParameters(thP,sys,par);
    smc.faPF(data, thSys, par)
    pSamp[kk] = smc.ll;
    
    # Fit the GP
    thPart      = th[0:kk+1,:];
    pSampPart   = pSamp[0:kk+1];
    m = GPy.models.GPRegression(thPart,pSampPart,kernel)
    
    # Find the maximum expected value
    Mup, ys2, up95, lo95 = m.predict( np.array(thPart) )
    mumax[kk] = np.max(Mup);
    
    # Compute the next point in which to sample the posterior
    x, fmin, ierror = solve(EIeval,l,u,user_data=(m,mumax[kk],par.epsilonEI),maxf=1000,maxT=1000);
    
    # Set the new point and save the estimate of the EI
    th[kk+1,:] = x;
    EI[kk+1] = fmin;
    
    tmp2, tmp3, ierror = solve(MUeval,l,u,user_data=m,algmethod=1);
    thhat[kk,:] = tmp2;
    llmax[kk,:] = tmp3;
    
    user_data = ( m,mumax[kk],par.epsilonEI)
    Mup, ys2, up95, lo95 = m.predict( np.array(grid.reshape((len(grid),1))) )
    
    for ii in range(0,len(grid)):    
        gridA[ii,kk], tmp = EIeval( np.array(grid[ii]), user_data )
    
    gridM[:,kk] = Mup[:,0];
    gridL[:,kk] = lo95[:,0];
    gridU[:,kk] = up95[:,0];
    
    print kk;

# Export data to R for plotting
out = hstack( ( gridM, gridL, gridU, gridA  ) )
pandas.DataFrame(out).to_csv("ch4-example-gpo-garchinference-ei.csv");

# Export data to R for plotting
out = hstack( ( th.transpose(), pSamp.transpose()  ) )
pandas.DataFrame(out).to_csv("ch4-example-gpo-garchinference-ei2.csv");

########################################################################
# GPO using the PI rule
########################################################################

# Preallocate the information needed
th     = np.zeros((par.maxIter+1,par.nPars))
EI     = np.zeros((par.maxIter+1,1))
mumax  = np.zeros((par.maxIter,1))
pSamp  = np.zeros((par.maxIter,1))
thhat  = np.zeros((par.maxIter,par.nPars))
llmax  = np.zeros((par.maxIter,1))
grid  = np.arange(0.01,0.26,0.01);
gridM = np.zeros((len(grid),par.maxIter));
gridL = np.zeros((len(grid),par.maxIter));
gridU = np.zeros((len(grid),par.maxIter));
gridA = np.zeros((len(grid),par.maxIter));

# Main GPO-loop
for kk in range(par.preIter-1,par.maxIter):

    # Sample the log-likelihood
    thP = (th[kk,:], sys.par[1], sys.par[2], sys.par[3]);
    thSys.storeParameters(thP,sys,par);
    smc.faPF(data, thSys, par)
    pSamp[kk] = smc.ll;
    
    # Fit the GP
    thPart      = th[0:kk+1,:];
    pSampPart   = pSamp[0:kk+1];
    m = GPy.models.GPRegression(thPart,pSampPart,kernel)
    
    # Find the maximum expected value
    Mup, ys2, up95, lo95 = m.predict( np.array(thPart) )
    mumax[kk] = np.max(Mup);
    
    # Compute the next point in which to sample the posterior
    x, fmin, ierror = solve(EIeval,l,u,user_data=(m,mumax[kk],par.epsilonEI),maxf=1000,maxT=1000);
    
    # Set the new point and save the estimate of the EI
    th[kk+1,:] = x;
    EI[kk+1] = fmin;
    
    tmp2, tmp3, ierror = solve(MUeval,l,u,user_data=m,algmethod=1);
    thhat[kk,:] = tmp2;
    llmax[kk,:] = tmp3;
    
    user_data = ( m,mumax[kk],par.epsilonEI)
    Mup, ys2, up95, lo95 = m.predict( np.array(grid.reshape((len(grid),1))) )
    
    for ii in range(0,len(grid)):    
        gridA[ii,kk], tmp = PIeval( np.array(grid[ii]), user_data )
    
    gridM[:,kk] = Mup[:,0];
    gridL[:,kk] = lo95[:,0];
    gridU[:,kk] = up95[:,0];
    
    print kk;

# Export data to R for plotting
out = hstack( ( gridM, gridL, gridU, gridA  ) )
pandas.DataFrame(out).to_csv("ch4-example-gpo-garchinference-pi.csv");

# Export data to R for plotting
out = hstack( ( th.transpose(), pSamp.transpose()  ) )
pandas.DataFrame(out).to_csv("ch4-example-gpo-garchinference-pi2.csv");

########################################################################
# GPO using the UCB rule
########################################################################

# Preallocate the information needed
th     = np.zeros((par.maxIter+1,par.nPars))
EI     = np.zeros((par.maxIter+1,1))
mumax  = np.zeros((par.maxIter,1))
pSamp  = np.zeros((par.maxIter,1))
thhat  = np.zeros((par.maxIter,par.nPars))
llmax  = np.zeros((par.maxIter,1))
grid  = np.arange(0.01,0.26,0.01);
gridM = np.zeros((len(grid),par.maxIter));
gridL = np.zeros((len(grid),par.maxIter));
gridU = np.zeros((len(grid),par.maxIter));
gridA = np.zeros((len(grid),par.maxIter));

# Main GPO-loop
for kk in range(par.preIter-1,par.maxIter):

    # Sample the log-likelihood
    thP = (th[kk,:], sys.par[1], sys.par[2], sys.par[3]);
    thSys.storeParameters(thP,sys,par);
    smc.faPF(data, thSys, par)
    pSamp[kk] = smc.ll;
    
    # Fit the GP
    thPart      = th[0:kk+1,:];
    pSampPart   = pSamp[0:kk+1];
    m = GPy.models.GPRegression(thPart,pSampPart,kernel)
    
    # Find the maximum expected value
    Mup, ys2, up95, lo95 = m.predict( np.array(thPart) )
    mumax[kk] = np.max(Mup);
    
    # Compute the next point in which to sample the posterior
    x, fmin, ierror = solve(EIeval,l,u,user_data=(m,mumax[kk],par.epsilonEI),maxf=1000,maxT=1000);
    
    # Set the new point and save the estimate of the EI
    th[kk+1,:] = x;
    EI[kk+1] = fmin;
    
    tmp2, tmp3, ierror = solve(MUeval,l,u,user_data=m,algmethod=1);
    thhat[kk,:] = tmp2;
    llmax[kk,:] = tmp3;
    
    user_data = ( m,mumax[kk],1.96)
    Mup, ys2, up95, lo95 = m.predict( np.array(grid.reshape((len(grid),1))) )
    
    for ii in range(0,len(grid)):    
        gridA[ii,kk], tmp = UCBeval( np.array(grid[ii]), user_data )
    
    gridM[:,kk] = Mup[:,0];
    gridL[:,kk] = lo95[:,0];
    gridU[:,kk] = up95[:,0];
    
    print kk;

# Export data to R for plotting
out = hstack( ( gridM, gridL, gridU, gridA  ) )
pandas.DataFrame(out).to_csv("ch4-example-gpo-garchinference-ucb.csv");

# Export data to R for plotting
out = hstack( ( th.transpose(), pSamp.transpose()  ) )
pandas.DataFrame(out).to_csv("ch4-example-gpo-garchinference-ucb2.csv");

########################################################################
# End of file
########################################################################